automation/generate_ami.sh

```diff
@@ -0,0 +1,181 @@
+#!/bin/bash
+
+set -e
+
+IMAGE="$IMAGE"
+
+AMI_NAME=$AMI_NAME
+AMI_ARCHITECTURE=${AMI_ARCHITECTURE:-x86_64}
+AMI_ROOT_DEVICE_NAME=${AMI_ROOT_DEVICE_NAME:-/dev/sda1}
+AMI_EBS_DELETE_ON_TERMINATION=${AMI_EBS_DELETE_ON_TERMINATION:-true}
+AMI_EBS_VOLUME_SIZE=${AMI_EBS_VOLUME_SIZE:-8}
+AMI_EBS_VOLUME_TYPE=${AMI_EBS_VOLUME_TYPE:-gp2}
+
+BALENA_PRELOAD_APP=${BALENA_PRELOAD_APP:-balena-os-config-preload-amd64}
+IMPORT_SNAPSHOT_TIMEOUT_MINS=${IMPORT_SNAPSHOT_TIMEOUT_MINS:-10}
+BOOT_PARTITION=""
+
+ensure_all_env_variables_are_set() {
+    local env_not_set=""
+    local env_variables="AWS_ACCESS_KEY_ID
+                         AWS_SECRET_ACCESS_KEY
+                         AWS_DEFAULT_REGION
+                         S3_BUCKET
+                         IMAGE
+                         AMI_NAME
+                         BALENARC_BALENA_URL
+                         BALENACLI_TOKEN
+                         PRELOAD_SSH_PUBKEY"
+
+    for env in $env_variables; do
+        [[ -z "${!env}" ]] && echo "ERROR: Missing env variable: $env" && env_not_set=true
+    done
+
+    if [[ "$env_not_set" ]]; then exit 1; fi
+}
+
+
+mount_cleanup() {
+    [[ -n "$BOOT_PARTITION" ]]              && \
+        echo "* Unmounting boot partition"  && \
+        sudo umount $BOOT_PARTITION         && \
+        rmdir $BOOT_PARTITION               && \
+        BOOT_PARTITION=""
+}
+
+
+
+mount_boot_partition() {
+    local img=$1
+    local __retvalue=$2
+
+    local sector_size
+    local partition_offset
+    local boot_partition_mountpoint
+
+    sector_size=$(fdisk -l "$img" | sed -n "s|Sector\ssize.*:\s\([0-9]\+\)\s.*$|\1|p")
+    partition_offset=$(fdisk -l "$img" | sed -n "s|${img}[0-9]\s\+\*\s\+\([0-9]\+\)\s\+.*$|\1|p")
+    boot_partition_mountpoint=$(mktemp -d)
+    sudo mount -o loop,offset=$((sector_size * partition_offset)) "$img" "$boot_partition_mountpoint"
+
+    echo "* Boot partition mounted on $boot_partition_mountpoint"
+    eval $__retvalue="'$boot_partition_mountpoint'"
+}
+
+deploy_preload_app_to_image() {
+
+    local image=$1
+
+    echo "* Adding the preload app"
+    balena preload                              \
+        --debug \
+        --app $BALENA_PRELOAD_APP               \
+        --commit current                        \
+        $image
+}
+
+add_ssh_key_to_boot_partition() {
+    local public_key=$1
+
+    echo "* Adding the preload public key"
+    cp ${BOOT_PARTITION}/config.json /tmp/.config.json
+    cat /tmp/.config.json | jq --arg keys "${public_key}" '. + {os: {sshKeys: [$keys]}}' | jq -r . > /tmp/.config.json.2
```

Comment:
You don't need two tmp files as jq won't modify the original. You can probably just:
```
jq --arg keys "${public_key}" '. + {os: {sshKeys: [$keys]}}' ${BOOT_PARTITION}/config.json > /tmp/.config.json
```

Reply:
That's the first thing I tried and you end up with an unmodified config.json...

Reply:
Not sure what you mean, I had tried this too:

Starting with:
```
root@balena:~# cat /mnt/boot/config.json | jq . 
{
  "apiEndpoint": "https://api.balena-cloud.com",
  "appUpdatePollInterval": 900000,
  "applicationId": 1602608,
  "deltaEndpoint": "https://delta.balena-cloud.com",
  "deviceApiKey": "7933c7caced5bd29998a5bb82001efdb",
  "deviceApiKeys": {
    "api.balena-cloud.com": "7933c7caced5bd29998a5bb82001efdb"
  },
  "deviceId": 4047471,
  "deviceType": "raspberrypi4-64",
  "hostappExtensions": "aggurio/ikheaders-hostext:uprobes aggurio/bpf-hostext:uprobes",
  "hostname": "balena",
  "listenPort": 48484,
  "localMode": true,
  "mixpanelToken": "9ef939ea64cb6cd8bbc96af72345d70d",
  "persistentLogging": false,
  "registered_at": 1604422306498,
  "registryEndpoint": "registry2.balena-cloud.com",
  "userId": 136580,
  "uuid": "dda20685f6cf6c82cb0dea31b77ef9da",
  "vpnEndpoint": "vpn.balena-cloud.com",
  "vpnPort": 443
}
```
```
BOOT_PARTITION=/mnt/boot
public_key=mykey

jq --arg keys "${public_key}" '. + {os: {sshKeys: [$keys]}}' ${BOOT_PARTITION}/config.json > /tmp/.config.json
```
The result is:
```
root@balena:~# cat /tmp/.config.json | jq .
{
  "apiEndpoint": "https://api.balena-cloud.com",
  "appUpdatePollInterval": 900000,
  "applicationId": 1602608,
  "deltaEndpoint": "https://delta.balena-cloud.com",
  "deviceApiKey": "7933c7caced5bd29998a5bb82001efdb",
  "deviceApiKeys": {
    "api.balena-cloud.com": "7933c7caced5bd29998a5bb82001efdb"
  },
  "deviceId": 4047471,
  "deviceType": "raspberrypi4-64",
  "hostappExtensions": "aggurio/ikheaders-hostext:uprobes aggurio/bpf-hostext:uprobes",
  "hostname": "balena",
  "listenPort": 48484,
  "localMode": true,
  "mixpanelToken": "9ef939ea64cb6cd8bbc96af72345d70d",
  "persistentLogging": false,
  "registered_at": 1604422306498,
  "registryEndpoint": "registry2.balena-cloud.com",
  "userId": 136580,
  "uuid": "dda20685f6cf6c82cb0dea31b77ef9da",
  "vpnEndpoint": "vpn.balena-cloud.com",
  "vpnPort": 443,
  "os": {
    "sshKeys": [
      "mykey"
    ]
  }
}

```

Reply:
It turns out that I needed this code to make it work **on my machine**. For some time I couldn't make the `balena preload` step work in a container, and therefore I had to test locally. For that, I needed this code (sudo, etc). Once I made everything work in a container, this could have been simplified, but I just forgot...

Fixed. Thanks for the comments :blush: